### PR TITLE
[Do Not Merge, need carbon-platform-integration 4.5.0 ] Introducing deployer suite listeners and jdbcGroup test suite

### PR DESCRIPTION
**Pull request by madumalt**

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityListenerConstants.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    public static final String TOMCAT_APPLICATIONS_LOCATIONS_PARAMETER = "tomcatWebAppsLocations";
+    public static final String TOMCAT_APPLICATIONS_URLS_PARAMETER = "tomcatWebAppsUrls";
+    public static final String PARAMETER_SEPARATOR = ",";
+    public static final int DEFAULT_TOMCAT_PORT = 8049;
+    public static final String ADDITIONAL_IS_COUNT_PARAMETER = "additionalISCount";
+    public static final String ADDITIONAL_IS_PORT_OFFSETS_PARAMETER = "additionalSPortOffsets";
+    public static final String ADDITIONAL_IS_PROUDUCT_GROUP_NAMES_PARAMETER = "additionalISProductGroupNames";
+    public static final String ADDITIONAL_IS_INSTANCE_NAMES_PARAMETER = "additionalISInstanceNames";
+    public static final String ADDITIONAL_IS_TEST_USER_MODES_PARAMETER = "additionalISTestUserModes";
+    public static final String ADDITIONAL_IS_PARAMETER_VALUE_SEPARATOR = ",";
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityListenerConstants.java]: remove this.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityPrimaryISDeployer.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                serverManager = new TestServerManager(automationContext, null, primayISParameters);
+            } catch (XPathExpressionException ex) {
+                throw new AutomationFrameworkException("Error while initializing automation context.", ex);
+            }
+
+            //Start the server if the execution environment is STANDALONE.
+            if (executionEnvironment.equalsIgnoreCase(ExecutionEnvironment.STANDALONE.name())) {
+                String carbonHome = serverManager.startServer(primayISParameters.get(IdentityListenerConstants.CONF_LOCATION));
+                System.setProperty(ExtensionConstants.CARBON_HOME, carbonHome);
+            }
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityPrimaryISDeployer.java]: what if not STANDALONE

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityPrimaryISDeployer.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                        ExtensionConstants.DEFAULT_CARBON_PORT_OFFSET));
+            }
+
+            // Initializing automation context.
+            try {
+                automationContext = new AutomationContext();
+                executionEnvironment = automationContext.getConfigurationValue(ContextXpathConstants.EXECUTION_ENVIRONMENT);
+                serverManager = new TestServerManager(automationContext, null, primayISParameters);
+            } catch (XPathExpressionException ex) {
+                throw new AutomationFrameworkException("Error while initializing automation context.", ex);
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityPrimaryISDeployer.java]: handle Exceptions

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityPrimaryISDeployer.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        // If the confLocation is empty then it implies that there are no config changes required
+        // for primaryIS deployment.
+        String relativeConfLocation = iSuite.getParameter(IdentityListenerConstants.CONF_LOCATION_PARAMETER);
+        if(StringUtils.isNotEmpty(relativeConfLocation)){
+            try {
+                primayISParameters.put(IdentityListenerConstants.CONF_LOCATION,
+                        FrameworkExtensionUtils.getResourceLocation(relativeConfLocation));
+            } catch (AutomationFrameworkException e) {
+                e.printStackTrace();
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityPrimaryISDeployer.java]: remove

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityServersDeployer.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+            // Number of portOffsets, instanceNames, and testUserModes should be equal to the requiredIScount.
+            if (ISsCount !=  portOffsets.split(IdentityListenerConstants.ADDITIONAL_IS_PARAMETER_VALUE_SEPARATOR)
+                    .length ||
+                    ISsCount != instanceNames.split(IdentityListenerConstants.ADDITIONAL_IS_PARAMETER_VALUE_SEPARATOR)
+                            .length ||
+                    ISsCount != testUserModes.split(IdentityListenerConstants.ADDITIONAL_IS_PARAMETER_VALUE_SEPARATOR)
+                    .length ) {
+                handleException("Incorrect additional ISs deployment parameters.");
+            }
+        } catch (Exception ex) {
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityServersDeployer.java]: NumberFormatException

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityServersDeployer.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        String productGroupNames = iSuite.getParameter(
+                IdentityListenerConstants.ADDITIONAL_IS_PROUDUCT_GROUP_NAMES_PARAMETER);
+        String instanceNames = iSuite.getParameter(IdentityListenerConstants.ADDITIONAL_IS_INSTANCE_NAMES_PARAMETER);
+        String testUserModes = iSuite.getParameter(IdentityListenerConstants.ADDITIONAL_IS_TEST_USER_MODES_PARAMETER);
+        // Validate additional ISs deployment parameters.
+        try{
+            // requiredISCount parameter value need to be a number.
+            int ISsCount = Integer.parseInt(requiredISCount);
+            // Number of portOffsets, instanceNames, and testUserModes should be equal to the requiredIScount.
+            if (ISsCount !=  portOffsets.split(IdentityListenerConstants.ADDITIONAL_IS_PARAMETER_VALUE_SEPARATOR)
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityServersDeployer.java]: move to a private method

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityServersDeployer.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        additionalISParameters.put(IdentityListenerConstants.ADDITIONAL_IS_PORT_OFFSETS_PARAMETER, portOffsets);
+        additionalISParameters.put(IdentityListenerConstants.ADDITIONAL_IS_PROUDUCT_GROUP_NAMES_PARAMETER,
+                productGroupNames);
+        additionalISParameters.put(IdentityListenerConstants.ADDITIONAL_IS_INSTANCE_NAMES_PARAMETER, instanceNames);
+        additionalISParameters.put(IdentityListenerConstants.ADDITIONAL_IS_TEST_USER_MODES_PARAMETER, testUserModes);
+    }
+
+    private IdentityServerParamsBean[] getIdentityServersParams(){
+
+        int ISsCount = Integer.parseInt(additionalISParameters.get(
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityServersDeployer.java]: think of using list

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityTomcatDeployer.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        try {
+            tomcat.start();
+        } catch (LifecycleException ex) {
+            throw new AutomationFrameworkException("Error occurred while starting tomcat.", ex);
+        }
+    }
+
+    private static void handleException(String msg, Exception e) {
```
Comment (madumalt) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/listeners/IdentityTomcatDeployer.java]: rename to imply the purpose

**Comment by JKAUSHALYA** [comment]
@madumalt Can we merge this ?

**Comment by isharak** [comment]
Close since this is no complete.

**Pull request closed without merging**